app/models/order.rb

```diff
@@ -0,0 +1,76 @@
+class Order < ApplicationRecord
+  include AASM
+
+  after_create :generate_order_number
+
+  has_one :coupon
+  has_one :cart
+  has_one :shipping_address, as: :addressable, dependent: :destroy
+  has_one :billing_address, as: :addressable, dependent: :destroy
+  has_many :order_items
+  belongs_to :user
+  belongs_to :delivery
+
+  aasm do
+    state :pending, initial: true
+    state :waiting_for_processing
+    state :in_progress
+    state :in_delivery
+    state :delivered
+    state :cancelled
+
+    event :waiting_processing do
+      transitions from: :pending, to: :waiting_for_processing
+    end
+
+    event :progress do
+      transitions from: :waiting_for_processing, to: :in_progress
+    end
+
+    event :delivered do
+      transitions from: :in_progress, to: :in_delivery
+    end
+
+    event :complete do
+      transitions from: :in_delivery, to: :delivered
+    end
+
+    event :cancelled do
+      transitions from: Order.aasm.states.map(&:name) - [:cancelled], to: :cancelled
+    end
+  end
+
+  def self.with_filter(filter)
```

Comment:
your filter values is equal to aasm_state, so you don't need to use case switch. You can just check if filter value in available states list.